src/main/java/org/terracotta/offheapstore/disk/storage/FileBackedStorageEngine.java

```diff
@@ -526,62 +480,36 @@ public void bind(Owner m) {
-        throw new RuntimeException(e);
-      } catch (OutOfMemoryError e) {
-        LOGGER.error("Failed to allocate direct buffer for FileChannel read.  "
-                + "Consider increasing the -XX:MaxDirectMemorySize property to "
-                + "allow enough space for the FileChannel transfer buffers");
-        throw e;
+      long position = filePosition + address;
+      FileWriteTask pending = pendingWrites.get(position);
+      if (pending == null) {
+        int keyLength = readBuffer(position + KEY_LENGTH_OFFSET, 4).getInt(0);
+        return readBuffer(position + KEY_DATA_OFFSET, keyLength);
+      } else {
+        return pending.getKeyBuffer();
       }
     }
 
     protected int readPojoHash(long address) {
-      try {
-        long position = filePosition + address;
-        FileWriteTask pending = pendingWrites.get(position);
-        if (pending == null) {
-          return readIntFromChannel(position + KEY_HASH_OFFSET);
-        } else {
-          return pending.pojoHash;
-        }
-      } catch (IOException e) {
-        throw new RuntimeException(e);
-      } catch (OutOfMemoryError e) {
-        LOGGER.error("Failed to allocate direct buffer for FileChannel read.  "
-                + "Consider increasing the -XX:MaxDirectMemorySize property to "
-                + "allow enough space for the FileChannel transfer buffers");
-        throw e;
+      long position = filePosition + address;
+      FileWriteTask pending = pendingWrites.get(position);
+      if (pending == null) {
+        return readBuffer(position + KEY_HASH_OFFSET, 4).getInt(0);
+      } else {
+        return pending.pojoHash;
       }
     }
 
     ByteBuffer readValueBuffer(long address) {
-      try {
-        long position = filePosition + address;
-        FileWriteTask pending = pendingWrites.get(position);
-        if (pending == null) {
-          int keyLength = readIntFromChannel(position + KEY_LENGTH_OFFSET);
-          int valueLength = readIntFromChannel(position + VALUE_LENGTH_OFFSET);
-          return readBuffer(position + keyLength + KEY_DATA_OFFSET, valueLength);
-        } else {
-          return pending.getValueBuffer();
-        }
-      } catch (IOException e) {
-        throw new RuntimeException(e);
-      } catch (OutOfMemoryError e) {
-        LOGGER.error("Failed to allocate direct buffer for FileChannel read.  "
-                + "Consider increasing the -XX:MaxDirectMemorySize property to "
-                + "allow enough space for the FileChannel transfer buffers");
-        throw e;
+      long position = filePosition + address;
+      FileWriteTask pending = pendingWrites.get(position);
+      if (pending == null) {
+        ByteBuffer lengths = readBuffer(position + KEY_LENGTH_OFFSET, 8);
+        int keyLength = lengths.getInt(0);
+        int valueLength = lengths.getInt(VALUE_LENGTH_OFFSET - KEY_LENGTH_OFFSET);
+        return readBuffer(position + keyLength + KEY_DATA_OFFSET, valueLength);
```

Comment:
This unnerves me a bit ... the original code read two ints each from a location based on position.  This code reads two ints with the first based on the original position and the second based not on position but on position of the first field -- an assumption is made about the relationship between the two fields that is not consistent with the original code -- `lengths` only contains 8 bytes of data ... what happens if the key and value length fields are not adjacent?

Reply:
Never trust git... this is a rebasing induced mistake. I'll do the rebase more carefully and it should clean up.